cpp/src/spatial/hausdorff.cu

```diff
@@ -26,89 +26,78 @@
 
 namespace {
 
-const uint32_t NUM_THREADS = 1024;
-
 template<typename T>
 constexpr auto magnitude_squared(T a, T b) {
     return a * a + b * b;
 }
 
 template <typename T>
 __global__ void kernel_hausdorff(int num_spaces,
-                                      T const* xs,
-                                      T const* ys,
-                                      cudf::size_type* space_offsets,
-                                      T* results)
+                                 T const* xs,
+                                 T const* ys,
+                                 cudf::size_type* space_offsets,
+                                 T* results)
 {
     auto block_idx = blockIdx.y * gridDim.x + blockIdx.x;
-    auto num_space_pairs = num_spaces * num_spaces;
+    auto num_results = num_spaces * num_spaces;
 
-    // each block processes a single pair of spaces.
-    if (block_idx < num_space_pairs)
+    // each block processes a single result / pair of spaces
+    if (block_idx >= num_results)
```

Comment:
This approach reduces nesting and eliminates the cognitive overhead of a potential `else` block much further down. The original implementation launches one block per pair, so this check may be out of an abundance of caution, and not necessary. Is changing the algorithm in scope for this PR? What do you mean by strided loop? I did not see a clear approach to unwrap the inner loop here, since it is dependent on the length of two separate spaces (trajectories is a misnomer. Hausdorff distance compares two spaces, finite subsets of which may include trajectories, curves, mesh vertices, point clouds, etc).

Reply:
But it doesn't let you do a strided loop, which is the idiomatic (and usually correct) way to implement most CUDA kernels over variable size data...

Reply:
If we can flatten the loops and stride them independent of the length of the spaces, that would be ideal, but I think that's out of scope for this PR. I'll get the if statement changed back to the nested version.

Reply:
Leaving this here in case I somehow forget:

Flatten this algorithm in case I lose it:

```
// ============================================================================
// find Hausdorff distance between spaces A, B, and C
//  - distance is non-symetrical, so calculate both ways
//  - results[i][j] = hausdorff distance from space i to space j
//  - distance between a space and itself is always zero
//    thus the diagonal of the matrix will always be zeroes
//
A = [025]
B = [37]
C = [9]

This is a single-dimension version of the Hausdorff algorithm, but basically `|v2-v1|` can be replaced with `magnitude(ax-bx, ay-by)` for a 2d version... or I guess any number of dimensions?

// ============================================================================
// cartesian product of input spaces (ABC)
//
AA, AB, AC
BA, BB, BC
CA, CB, CC

// expand by the cartesian product of each pair of spaces
//
[025]x[025], [025]x[37], [025]x[9]
 [37]x[025],  [37]x[37],  [37]x[9]
  [9]x[025],   [9]x[37],   [9]x[9]

// group of spaces is indicated by the bars across the top and sides of the matrix ("---==" and "][").
//
//    ================  ==========  ====
//    ================  ==========  ====
// ][ (00), (02), (05), (03), (07), (09)
// ][
// ][ (20), (22), (25), (23), (27), (29)
// ][
// ][ (50), (52), (55), (53), (57), (59)
// 
// ][ (30), (32), (35), (33), (37), (39)
// ][
// ][ (70), (72), (75), (73), (77), (79)
//
// ][ (90), (92), (95), (93), (97), (99)

// ============================================================================
// find the minimum distance from each point in a space to any point in another space
// find the maximum distance of each resulting distance.
// NOTE: word this better.
//
//    --------===================   --------============   --------=====
//    --------===================   --------============   --------=====
// ][ max(min(|0-0|, |0-2|, |0-5|), max(min(|0-3|, |0-7|), max(min(|0-9|),
// ][     min(|2-0|, |2-2|, |2-5|),     min(|2-3|, |2-7|),     min(|2-9|),
// ][     min(|5-0|, |5-2|, |5-5|))     min(|5-3|, |5-7|))     min(|5-9|))
// 
// ][ max(min(|3-0|, |3-2|, |3-5|), max(min(|3-3|, |3-7|), max(min(|3-9|),
// ][     min(|7-0|, |7-2|, |7-5|))     min(|7-3|, |7-7|))     min(|7-9|))
// 
// ][ max(min(|9-0|, |9-2|, |9-5|)) max(min(|9-3|, |9-7|)) max(min(|9-9|))

//    ----=====  ----=====  ----=====
//    ----=====  ----=====  ----=====
// ][ max(|0-0|, max(|0-3|, max(|0-9|,
// ][     |2-2|,     |2-3|,     |2-9|,
// ][     |5-5|)     |5-3|)     |5-9|)
//
// ][ max(|3-2|, max(|3-3|, max(|3-9|,
// ][     |7-5|)     |7-7|)     |7-9|)
//
// ][ max(|9-5|) max(|9-7|) max(|9-9|)

//    ===== ===== =====
// ][ |0-0| |5-3| |0-9|
// ][ |7-5| |7-7| |3-9|
// ][ |9-5| |9-7| |9-9|

//    = = =
// ][ 0 2 9
// ][ 2 0 6
// ][ 4 2 0
```

Translating this to thrust could (hypothetically) be done by
1) compute cartesian product of the input points
2) convert each element in the product to a distance between said points.
3) reduce_by_key where the values are points, keys are the shape of the _first_ point in the cartesian pairs, and the reduce function is `min`.
4) reduce where the values are the min distances and the reduce function is `max`.

note: the lookup "shape of the _first_ point in the cartesian pairs" can be generated by scattering 1s in to an array length of the same length as the number of total points, where the scatter indices are the offsets of the groups. Then pass that through a scan.
`[4,2,3,1] => [0,0,0,0,1,1,2,2,2,3]`

As for the performance, I haven't calculated. This is a rough draft.